packages/web3-eth/src/eth_tx.ts

```diff
@@ -0,0 +1,315 @@
+
+	validateCustomChainInfo(transaction);
+	validateChainInfo(transaction);
+	validateGas(formattedTransaction);
+
+	if (
+		formattedTransaction.nonce === undefined ||
+		formattedTransaction.chainId === undefined ||
+		formattedTransaction.nonce.startsWith('-') ||
+		formattedTransaction.chainId.startsWith('-')
+	)
+		throw new InvalidNonceOrChainIdError({
+			nonce: transaction.nonce,
+			chainId: transaction.chainId,
+		});
+};
+
+export async function populateTransaction<
+	DesiredType extends ValidTypes,
+	ReturnType = ValidReturnTypes[DesiredType],
+>(
+	transaction: Transaction,
+	web3Context: Web3Context<EthExecutionAPI>,
+	desiredType: DesiredType,
+	privateKey?: HexString,
+	overrideMethod?: (
+		transaction: Transaction,
+		web3Context: Web3Context<EthExecutionAPI>,
+	) => PopulatedUnsignedTransaction<ReturnType>,
+): Promise<PopulatedUnsignedTransaction<ReturnType>> {
+	if (overrideMethod !== undefined) return overrideMethod(transaction, web3Context);
+
+	const populatedTransaction = { ...transaction };
+	const web3Eth = new Web3Eth(web3Context.currentProvider);
+
+	if (populatedTransaction.from === undefined) {
+		if (privateKey !== undefined) {
+			populatedTransaction.from = privateKeyToAddress(privateKey);
+		} else if (web3Context.defaultAccount !== null)
+			populatedTransaction.from = web3Context.defaultAccount;
+		// TODO Try to fill from using web3.eth.accounts.wallet
+	}
+
+	if (populatedTransaction.nonce === undefined) {
+		if (populatedTransaction.from === undefined) throw new UnableToPopulateNonceError();
+		populatedTransaction.nonce = await web3Eth.getTransactionCount(
+			populatedTransaction.from,
+			BlockTags.PENDING,
+		);
+	}
+
+	if (populatedTransaction.value === undefined) populatedTransaction.value = '0x';
+	if (populatedTransaction.data === undefined) populatedTransaction.data = '0x';
+	// TODO - Add default to Web3Context
+	if (populatedTransaction.chain === undefined) populatedTransaction.chain = 'mainnet';
+	// TODO - Add default to Web3Context
+	if (populatedTransaction.hardfork === undefined) populatedTransaction.hardfork = 'london';
+
+	if (populatedTransaction.chainId === undefined) {
+		if (populatedTransaction.common?.customChain.chainId === undefined) {
+			// TODO - web3Eth.getChainId not implemented
+			// populatedTransaction.chainId = await web3Eth.getChainId();
+		}
+	}
+
+	if (populatedTransaction.gas === undefined) {
+		if (populatedTransaction.gasLimit !== undefined)
+			populatedTransaction.gas = populatedTransaction.gasLimit;
+	}
+
+	if (populatedTransaction.gasLimit === undefined) {
+		if (populatedTransaction.gas !== undefined)
+			populatedTransaction.gasLimit = populatedTransaction.gas;
+	}
+
+	populatedTransaction.type = detectTransactionType(populatedTransaction);
+	// TODO - After web3Context.defaultTxType is implemented
+	if (populatedTransaction.type === undefined) populatedTransaction.type = '0x0'; // web3Context.defaultTxType;
+
+	const block = await web3Eth.getBlock();
```

Comment:
`const block = await web3Eth.getBlock();` should be moved after line#288 : `if (hexTxType === '0x2') {`

as `block.baseFeePerGas` is only used if tx is 0x2, little performance imp.